cardano-testnet/src/Testnet/Runtime.hs

```diff
@@ -218,97 +220,57 @@ startLedgerNewEpochStateLogging testnetRuntime tmpWorkspace = withFrozenCallStac
       H.evalIO $ appendFile logFile ""
       socketPath <- H.noteM $ H.sprocketSystemName <$> H.headM (poolSprockets testnetRuntime)
 
-      runInBackground
-        (do logFileContents <- IO.readFile logFile
-            let epochStateValues = epochStateBeforeAfterValues logFileContents
-                epochStateDiffs' = epochStateDiffs epochStateValues
-            Text.writeFile (logDir </> "ledger-epoch-state-diffs.log") epochStateDiffs'
-        )
-        (do void $ runExceptT $
-              foldEpochState
-                (configurationFile testnetRuntime)
-                (Api.File socketPath)
-                Api.QuickValidation
-                (EpochNo maxBound)
-                ()
-                (\epochState _ _ ->
-                    liftIO $ evalStateT (handler logFile epochState) 0
-                )
-        )
-
-
-      H.note_ $ "Started logging epoch states to: " <> logFile
+      _ <- runInBackground . runExceptT $ do
+        foldEpochState
+          (configurationFile testnetRuntime)
+          (Api.File socketPath)
+          Api.QuickValidation
+          (EpochNo maxBound)
+          Nothing
+          (handler logFile diffFile)
+
+      H.note_ $ "Started logging epoch states to: " <> logFile <> "\nEpoch state diffs are logged to: " <> diffFile
 
   where
-    handler :: FilePath -> AnyNewEpochState -> StateT Int IO LedgerStateCondition
-    handler outputFpHandle (AnyNewEpochState sbe nes) = do
-      handleException . liftIO $ do
-        appendFile outputFpHandle $ "#### BLOCK ####" <> "\n"
-        appendFile outputFpHandle $ BSC.unpack (shelleyBasedEraConstraints sbe $ encodePretty nes) <> "\n"
-        pure ConditionNotMet
+    handler :: FilePath -- ^ log file
+            -> FilePath -- ^ diff file
+            -> AnyNewEpochState
+            -> SlotNo
+            -> BlockNo
+            -> StateT (Maybe AnyNewEpochState) IO LedgerStateCondition
+    handler outputFp diffFp anes@(AnyNewEpochState !sbe !nes) _ (BlockNo blockNo) = handleException $ do
+      let prettyNes = shelleyBasedEraConstraints sbe (encodePretty nes)
+          blockLabel = "#### BLOCK " <> show blockNo <> " ####"
+      liftIO . BSC.appendFile outputFp $ BSC.unlines [BSC.pack blockLabel, prettyNes, ""]
+
+      -- store epoch state for logging of differences
+      mPrevEpochState <- state $ swap . (Just anes,)
```

Comment:
```suggestion
      mPrevEpochState <- get
      put (Just anes)
```
I mean, the other way is really clever, but it took me several minutes to understand it